File: clientJava/src/main/java/at/rtr/rmbt/client/Pretest.java
package at.rtr.rmbt.client;

import java.io.IOException;

/** Single-thread 2-second download pre-test to determine chunk size and thread counts. */
final class Pretest {

    private static final long PRETEST_DURATION_NS = 2_000_000_000L;
    private static final int  MIN_CHUNK           = 1024;
    private static final int  MAX_CHUNK           = 4 * 1024 * 1024;

    private static final double[] DL_THRESH = {100.0, 1.0, 0.0};
    private static final int[]    DL_COUNT  = {  5,   3,   1 };
    private static final double[] UL_THRESH = {150.0, 80.0, 30.0, 0.0};
    private static final int[]    UL_COUNT  = {  5,    3,    2,   1 };

    private Pretest() {}

    static PretestResult run(String addr, int port, boolean useTls, boolean noTlsVerify,
                             int protocol, String token, int maxThreads) throws IOException {
        System.out.println("\nPre-test: measuring baseline throughput...");

        try (RmbtConn conn = RmbtConn.connect(addr, port, useTls, noTlsVerify, protocol)) {
            conn.greeting(token);

            int serverMin = conn.chunkSizeMin;
            int serverMax = conn.chunkSizeMax;

            long tStart    = System.nanoTime();
            int  cs        = Math.max(serverMin, MIN_CHUNK);
            int  n         = 1;
            long lastBytes = 0;
            long lastNs    = 0;  // TIME from the last (largest) GETCHUNKS batch
            long rttNs     = 0;  // TIME of the first (tiny) batch ≈ round-trip time

            while (System.nanoTime() - tStart < PRETEST_DURATION_NS) {
                String accept = conn.readLine();
                if (!accept.contains("GETCHUNKS"))
                    throw new IOException("Pre-test: expected GETCHUNKS, got: " + accept);

                conn.writeLine("GETCHUNKS " + n + " " + cs);

                byte[] buf = new byte[cs];
                for (int i = 0; i < n; i++) conn.readExact(buf);
                lastBytes = (long)n * cs;

                conn.writeLine("OK");
                String timeLine = conn.readLine();
                try {
                    long t = Long.parseLong(timeLine.split("\\s+")[1]);
                    if (rttNs == 0) rttNs = t;  // first tiny batch ≈ RTT
                    lastNs = t;
                } catch (Exception ignored) {}

                if (n >= 8) {
                    cs = Math.min(cs * 2, Math.min(serverMax, MAX_CHUNK));
                    n  = 1;
                } else {
                    n *= 2;
                }
            }

            // TIME = transmission_time + RTT.  Subtract the RTT estimate (first tiny
            // batch, where transmission_time ≈ 0) to get actual throughput.
            long transferNs = Math.max(0, lastNs - rttNs);
            double bps = transferNs > 0 ? lastBytes / (transferNs / 1e9)
                       : lastNs   > 0  ? lastBytes / (lastNs    / 1e9)
                       : lastBytes / ((System.nanoTime() - tStart) / 1e9);
            double mbps = bps * 8.0 / 1e6;

            int ideal    = (int)(bps / 50.0);
            int rounded  = ((ideal + 512) / 1024) * 1024;
            int floor    = Math.max(serverMin, MIN_CHUNK);
            int cap      = Math.min(serverMax, MAX_CHUNK);
            int chunkSize = Math.max(floor, Math.min(cap, rounded == 0 ? floor : rounded));

            int dl = Math.min(maxThreads, threadsFor(mbps, DL_THRESH, DL_COUNT));
            int ul = Math.min(maxThreads, threadsFor(mbps, UL_THRESH, UL_COUNT));

            System.out.printf("  pre-test: %.1f Mbit/s → chunk=%d KiB  dl_threads=%d  ul_threads=%d%n",
                    mbps, chunkSize / 1024, dl, ul);

            conn.quit();
            return new PretestResult(chunkSize, dl, ul);
        }
    }

    private static int threadsFor(double mbps, double[] thresh, int[] count) {
        for (int i = 0; i < thresh.length; i++)
            if (mbps >= thresh[i]) return count[i];
        return 1;
    }
}

File: clientRust/src/pretest.rs
use anyhow::{bail, Result};
use std::time::{Duration, Instant};

use crate::connection::{RmbtConn, Protocol};

const PRETEST_DURATION: Duration = Duration::from_millis(2000);
const MIN_CHUNK: usize = 1024;
const MAX_CHUNK: usize = 4 * 1024 * 1024;

pub struct PretestResult {
    pub chunk_size: usize,
    pub dl_threads: usize,
    pub ul_threads: usize,
}

// Rows are (min_mbps, thread_count), sorted descending so the first match wins.
const DL_TABLE: &[(f64, usize)] = &[(100.0, 5), (1.0, 3), (0.0, 1)];
const UL_TABLE: &[(f64, usize)] = &[(150.0, 5), (80.0, 3), (30.0, 2), (0.0, 1)];

fn threads_for(mbps: f64, table: &[(f64, usize)]) -> usize {
    table.iter()
         .find(|&&(thresh, _)| mbps >= thresh)
         .map(|&(_, c)| c)
         .unwrap_or(1)
}

/// Run a 2-second single-thread download pre-test using GETCHUNKS with
/// exponential chunk-count and chunk-size progression.
/// Returns the optimal chunk size and suggested thread counts for each direction.
pub fn run_pretest(
    addr:          &str,
    port:          u16,
    use_tls:       bool,
    no_tls_verify: bool,
    protocol:      Protocol,
    token:         &str,
    max_threads:   usize,
) -> Result<PretestResult> {
    println!("\nPre-test: measuring baseline throughput…");

    let mut conn = RmbtConn::connect(addr, port, use_tls, no_tls_verify, protocol)?;
    conn.greeting(token)?;

    let server_min = conn.chunk_size_min;
    let server_max = conn.chunk_size_max;

    let t_start        = Instant::now();
    let mut cs         = server_min.max(MIN_CHUNK);
    let mut n          = 1usize;
    let mut last_bytes = 0u64;
    let mut last_ns    = 0u64;  // TIME from the last (largest) GETCHUNKS batch
    let mut rtt_ns     = 0u64;  // TIME of the first (tiny) batch ≈ round-trip time

    loop {
        if t_start.elapsed() >= PRETEST_DURATION { break; }

        let accept = conn.read_accept()?;
        if !accept.contains("GETCHUNKS") {
            bail!("pre-test: expected ACCEPT with GETCHUNKS, got: {accept}");
        }

        conn.write_line(&format!("GETCHUNKS {n} {cs}"))?;

        let mut buf = vec![0u8; cs];
        for _ in 0..n {
            conn.read_exact(&mut buf)?;
        }
        last_bytes = (n as u64) * (cs as u64);

        conn.write_line("OK")?;
        let time_line = conn.read_line()?;
        if let Ok(ns) = crate::tests::parse_time_ns(&time_line) {
            if rtt_ns == 0 { rtt_ns = ns; }  // first tiny batch ≈ RTT
            last_ns = ns;
        }

        // Exponential progression: double n until n≥8, then double chunk size.
        if n >= 8 {
            cs = (cs * 2).min(server_max.min(MAX_CHUNK));
            n = 1;
        } else {
            n *= 2;
        }
    }

    // TIME = transmission_time + RTT.  Subtract the RTT estimate (from the
    // first tiny batch where transmission time ≈ 0) to get actual throughput.
    let transfer_ns = last_ns.saturating_sub(rtt_ns);
    let bps = if transfer_ns > 0 {
        last_bytes as f64 / (transfer_ns as f64 / 1_000_000_000.0)
    } else if last_ns > 0 {
        last_bytes as f64 / (last_ns as f64 / 1_000_000_000.0)
    } else {
        last_bytes as f64 / t_start.elapsed().as_secs_f64()
    };
    let mbps = bps * 8.0 / 1_000_000.0;

    // Target 50 chunks/sec (1 chunk per 20 ms), rounded to the nearest KiB.
    let ideal     = (bps / 50.0) as usize;
    let rounded   = ((ideal + 512) / 1024) * 1024;
    let chunk_size = rounded
        .max(server_min.max(MIN_CHUNK))
        .min(server_max.min(MAX_CHUNK));

    let dl = threads_for(mbps, DL_TABLE).min(max_threads);
    let ul = threads_for(mbps, UL_TABLE).min(max_threads);

    println!(
        "  pre-test: {mbps:.1} Mbit/s → chunk={} KiB  dl_threads={dl}  ul_threads={ul}",
        chunk_size / 1024
    );

    conn.quit()?;
    Ok(PretestResult { chunk_size, dl_threads: dl, ul_threads: ul })
}
